### DISCUSSION TITLE
Regenerated PAT

### DISCUSSION THREAD

**Discussion by Franklin-Ngueiwoh** [community/community#33049] [New to GitHub] [answered]
My PAT has expired and I was promoted to regenerate one. However, after regenerating the PAT, I am unable to do git push and am equally unable to use the new PAT to reclone already existing repositories. How do I proceed to work locally and push without being asked for my password every time I need to git push?

**Comment by jfullstackdev** [comment]
do you know how to use GitHub Desktop ? you can do that without using Git, if say, that is you only developing the app, or maybe you are in a team but say, you are the only assigned in one module, so you don't need to pull the changes from the other modules

**Reply by Franklin-Ngueiwoh** [reply]
I actually have github desktop and VS code. I am however not permitted to use it for the duration fo the course I am taking - we are only permitted to use our provided sandboxes which is a virtual machine

**Comment by airtower-luna** [comment]
What happens is probably that your old PAT is cached *somewhere*. You'll have to update that storage, or clear it so you'll be asked again.

Where that cache might be depends on:

* Your OS/desktop environment.
* The software you use. In particular, does it call Git, or have its own Git implementation and auth (e.g. Eclipse does this)?
* Your Git configuration, for example you might have configured [Git credential storage](https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage) (or something did that for you). Some people also embed tokens in remote URLs (which is generally a bad idea).

**Reply by Franklin-Ngueiwoh** [reply]
I started an online course and we were assigned sandboxes (virtual machines on web terminals). I am guessing that is where I will have to look for where the old PAT is cached. Any guide on how to go about that?

**Reply by airtower-luna** [reply]
If you have a command line, run `git config --list` and see if any of the options mentioned in the credential storage docs are set.

Otherwise I can't really help because I don't know that web environment. You might want to ask whoever is offering the course about that.

**Reply by Franklin-Ngueiwoh** [reply]
When run git config --list it gives me:
user.email
user.name

**Reply by airtower-luna** [reply]
Then you should really ask whoever is responsible for that sandbox. It's odd that a web-based course would ask you for a PAT anyway, it'd make more sense for them to use a GitHub App or OAuth to get access to the relevant repositories in your account. :thinking:

**Reply by Franklin-Ngueiwoh** [reply]
Thanks for the advice. I have sent a message to our tutors. Hopefully, they will assist me.

**Reply by wisdom-jessey** [reply]
Please can you explain how you did  it

**Comment by Franklin-Ngueiwoh** [comment, marked as answer]
Thanks for all the help. I figured out how to solve it using git remote set-url --delete <name> <oldURL> to delete the old URL then git remote set-url --add <name> <newULR> to add the new URL
All good now

**Reply by airtower-luna** [reply]
That sounds like you actually put the PAT into the URL, which is less than advisable in terms of security. Or convenience, because as you noticed you have to update the URL every time you change your authentication. :sweat_smile:

**Reply by Franklin-Ngueiwoh** [reply]
I only noticed this security measure now and its deficiencies. Probably done that way because the sandboxes given to us are just for school purposes.

**Reply by airtower-luna** [reply]
Really, that's a lot more reason *not* to store the token on disk. Sandbox or not, the PAT gives push access to your repositories. So if someone manages to steal it they can probably do a lot of damage.

**Reply by Franklin-Ngueiwoh** [reply]
Thanks for the help. I will make sure it is well protected.
